src/System.Net.Http/src/System/Net/Http/HPack/DynamicTable.cs

```diff
@@ -0,0 +1,94 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+namespace System.Net.Http.HPack
+{
+    internal class DynamicTable
+    {
+        private HeaderField[] _buffer;
+        private int _maxSize;
+        private int _size;
+        private int _count;
+        private int _insertIndex;
+        private int _removeIndex;
+
+        public DynamicTable(int maxSize)
+        {
+            _buffer = new HeaderField[maxSize / HeaderField.RfcOverhead];
+            _maxSize = maxSize;
+        }
+
+        public int Count => _count;
+
+        public int Size => _size;
+
+        public int MaxSize => _maxSize;
+
+        public HeaderField this[int index]
+        {
+            get
+            {
+                if (index >= _count)
+                {
+                    throw new IndexOutOfRangeException();
+                }
```

Comment:
When would this condition be hit?  Should it be an assert instead, or can bad data from the server or a user cause this?

Reply:
Bad data from the server can cause this. E.g. they send a header ID that's invalid.

For what it's worth, the only place we call this (HPackDecoder.GetHeader) catches this exception and throws HPackDecodingException instead.